### PR TITLE
Fix handling of negative intervals

### PR DESCRIPTION
**Pull request by electrum**

### CODE REVIEW

**Comment by electrum** [comment]
Fixes #5448

File: presto-docs/src/main/sphinx/release/release-0.148.rst [hunk 1]
```diff
@@ -21,8 +21,10 @@ General Changes
   honored by node scheduler. This bug could stop the cluster from making further progress.
 * Fix handling of ``INTERVAL DAY TO SECOND`` type in REST API. Previously, intervals greater than
   ``2,147,483,647`` milliseconds (about ``24`` days) were returned as the wrong value.
+  Additionally, negative intervals were formatted incorrectly and could not be read.
 * Fix handling of ``INTERVAL YEAR TO MONTH`` type in REST API. Previously, intervals greater than
   ``2,147,483,647`` months were returned as the wrong value.
+  Additionally, negative intervals were formatted incorrectly and could not be read.
```
Comment (dain) [inline presto-docs/src/main/sphinx/release/release-0.148.rst]: did you mean `year to month intervals`

File: presto-docs/src/main/sphinx/release/release-0.148.rst [hunk 2]
```diff
@@ -21,8 +21,10 @@ General Changes
   honored by node scheduler. This bug could stop the cluster from making further progress.
 * Fix handling of ``INTERVAL DAY TO SECOND`` type in REST API. Previously, intervals greater than
   ``2,147,483,647`` milliseconds (about ``24`` days) were returned as the wrong value.
+  Additionally, negative intervals were formatted incorrectly and could not be read.
```
Comment (dain) [inline presto-docs/src/main/sphinx/release/release-0.148.rst]: did you mean `day to second intervals`

File: presto-client/src/main/java/com/facebook/presto/client/IntervalDayTime.java [hunk 3]
```diff
@@ -55,4 +63,30 @@ public static String formatMillis(long millis)
 
         return format("%s%d %02d:%02d:%02d.%03d", sign, day, hour, minute, second, millis);
     }
+
+    public static long parseMillis(String value)
+    {
+        if (value.equals(LONG_MIN_VALUE)) {
```
Comment (dain) [inline presto-client/src/main/java/com/facebook/presto/client/IntervalDayTime.java]: So the parser only handles the exact format of the format with no extra whitespace... maybe document that

**Comment by dain** [comment]
Other than the build failures, looks good

File: presto-docs/src/main/sphinx/release/release-0.148.rst (hunk 1 above)
Comment (electrum) [inline presto-docs/src/main/sphinx/release/release-0.148.rst]: No, this is another sentence as part of the same list item, so I don't think repeating the full type name is necessary.

Or should I split them out into a single, separate item?

File: presto-client/src/main/java/com/facebook/presto/client/IntervalDayTime.java (hunk 3 above)
Comment (electrum) [inline presto-client/src/main/java/com/facebook/presto/client/IntervalDayTime.java]: Good point, I'll fix it to handle whitespace

Comment (electrum) [inline presto-client/src/main/java/com/facebook/presto/client/IntervalDayTime.java]: Actually, I'll just add documentation

Comment (electrum) [inline presto-client/src/main/java/com/facebook/presto/client/IntervalDayTime.java]: It turns out the parser used for SQL doesn't handle whitespace, either. Should it?

File: presto-client/src/main/java/com/facebook/presto/client/IntervalDayTime.java [hunk 4]
```diff
@@ -34,11 +36,17 @@ private IntervalDayTime() {}
...
+        try {
+            long value = millis;
+            value = addExact(value, multiplyExact(day, MILLIS_IN_DAY));
+            value = addExact(value, multiplyExact(hour, MILLIS_IN_HOUR));
+            value = addExact(value, multiplyExact(minute, MILLIS_IN_MINUTE));
+            value = addExact(value, multiplyExact(second, MILLIS_IN_SECOND));
+            return value;
+        }
+        catch (ArithmeticException e) {
+            throw new IllegalArgumentException(e);
```
Comment (dain) [inline presto-client/src/main/java/com/facebook/presto/client/IntervalDayTime.java]: IIRC there is a specific Presto exception for this

**Comment by dain** [comment]
Assuming the tests pass, looks good

**Pull request closed without merging**